scripts/linux/nestedAgentLock.sh

```diff
@@ -0,0 +1,205 @@
+        elif [ $save_next_arg -eq 2 ]; then
+            AGENT_GROUP="$arg"
+            save_next_arg=0
+        elif [ $save_next_arg -eq 3 ]; then
+            BUILD_ID="$arg"
+            save_next_arg=0
+        else
+            case "$arg" in
+                "-h" ) usage;;
+                "-p" ) save_next_arg=1;;
+                "-a" ) save_next_arg=2;;
+                "-b" ) save_next_arg=3;;
+                * ) usage;;
+            esac
+        fi
+    done
+
+    if [[ -z ${PAT} ]]; then
+        echo "Personal Access Token is a required parameter"
+        print_help_and_exit
+    fi
+
+    if [[ -z ${AGENT_GROUP} ]]; then
+        echo "Agent Group is a required parameter."
+        print_help_and_exit
+    fi
+
+    if [[ -z ${BUILD_ID} ]]; then
+        echo "Build id is a required parameter."
+        print_help_and_exit
+    fi
+}
+
+###############################################################################
+# Main Script Execution
+###############################################################################
+
+function set_devops_output_vars() {
+    agents=("$@")
+    outputAgentNames=(l5AgentName l4AgentName l3AgentName)
+
+    echo "Setting devops vars for agent names. Needed for future unlock."
+
+    for i in "${!agents[@]}"; do
+        agentId="${agents[$i]}"
+
+        agentCapabilities=$(curl -s -u :$PAT --request GET "https://dev.azure.com/msazure/_apis/distributedtask/pools/$POOL_ID/agents/$agentId?includeCapabilities=true&api-version=$API_VER")
+        agentName=$(echo $agentCapabilities | jq '.systemCapabilities."Agent.Name"' | tr -d '[], "')
+
+        echo "Setting devops var for agent name: $agentName"
+        echo "##vso[task.setvariable variable=${outputAgentNames[$i]};isOutput=true]$agentName"
+    done
+}
+
+function update_capabilities() {
+    agentId=$1
+    newAgentUserCapabilities=$2
+
+    # Update the user capability on the agent pool for this agent
+    responseCapabilities=$(curl -s -u :$PAT \
+--request PUT "https://msazure.visualstudio.com/_apis/distributedtask/pools/$POOL_ID/agents/$agentId/usercapabilities" \
+-H "Content-Type:application/json" \
+-H "Accept: application/json;api-version=5.0;" \
+--data @<(cat <<EOF
+$newAgentUserCapabilities
+EOF
+))
+    # Validate the capability update was successful
+    responseUserCapabilities=$(echo $responseCapabilities | jq '.userCapabilities')
+
+    if [ "$responseUserCapabilities" != "$newAgentUserCapabilities" ]
+    then
+        echo "Capabilities were not updated properly. This will be retried."
+    fi
+}
+
+process_args $@
+
+# Install pre-requisite 'jq'
+CMD=jq
+echo "Validating jq"
+if ! command -v $CMD &>/dev/null; then
+    echo "Command '$CMD' not found, Installing '$CMD'"
+    sudo add-apt-repository universe
+    sudo apt-get update
+    sudo apt-get install -y $CMD
+fi
+echo "Done validating jq"
+
+startSeconds=$((SECONDS))
+endSeconds=$((SECONDS + $TIMEOUT_SECONDS))
+while true && [ $((SECONDS)) -lt $endSeconds ]; do
```

Comment:
The body of this while loop is well past my personal threshold for cleanup. Never wrote a bash script with complicated logic before, but I assumed I could clean it up using functions with return values.

Unfortunately bash functions don't work like expected. The common pattern is to capture the stdout of the function as the return value. But this is not good if we want to use any logging. Any suggestions on what to do?
https://stackoverflow.com/questions/17336915/return-value-in-a-bash-function

Reply:
I tried to make it better despite the limitation.

Reply:
I see what you're trying to do. Here are some ways you can do it: 
https://www.linuxjournal.com/content/return-values-bash-functions

Note: you also have an option to direct set it to the pipeline variable i.e.
`echo "##vso[task.setvariable variable=VERSION;]$VERSION" `

Also Could you make this and exponential back off?

Reply:
Yeah I can explore exponential backoff